core/jvm/src/main/scala/scalapb/zio_grpc/ZCall.scala

```diff
@@ -0,0 +1,39 @@
+package scalapb.zio_grpc
+
+import io.grpc.Status
+import scalapb.zio_grpc.ZCall.ReadyPromise
+import zio.{Promise, Ref, UIO, ZIO}
+
+object ZCall {
+  type ReadyPromise = Ref[Option[Promise[Nothing, Unit]]]
+}
+
+trait ZCall[-R, A] extends Any {
+
+  def request(n: Int): ZIO[R, Status, Unit]
+
+  def sendMessage(message: A): ZIO[R, Status, Unit]
+
+  def sendMessageWhenReady(message: A): ZIO[R, Status, Unit] =
+    ZIO.ifM(isReady)(
+      sendMessage(message),
+      Promise.make[Nothing, Unit].flatMap { promise =>
+        readyPromise.set(Some(promise)) *> ZIO.ifM(isReady)(
```

Comment:
should we assert on whether the promise is already set?

Reply:
With respect to fs2-grpc, fs2.Stream has a pauseWhen method, so I was thinking if `stream.pauseWhen(Stream.repeatEval(notReady)).evalMap(sendStream)` would accomplish the same. Not sure if that works, though, but simpler than dealing with onReady double-check callback from listener.

Reply:
Good question!  
Now all usages of `sendMessageWhenReady` are serial, but in case of concurrent calls one can rewrite others promise.
But it can be hard to test such uncommon fail (because of race conditions), so end user of the library can face this problem.
Maybe we should protect `sendMessageWhenReady` with `Semaphore`?
Or if concurrent calls to `sendMessageWhenReady` are allowed should we may replace `readyPromise` with a `Queue` for synchronisation between multiple `send` and `ready` calls?